File: topk.pl
use strict;
use warnings;

# Returns the k most frequent elements in the given array
sub k_most_frequent {
  my ($nums, $k) = @_;

  # Return an empty array if the input array is empty
  return [] if @$nums == 0;

  # Return the entire array if k is greater than the number of elements
  return @$nums if $k > @$nums;

  # Use a hash to count the frequency of each element
  my %count;
  foreach my $num (@$nums) {
    $count{$num}++;
  }

  # Create a min-heap of elements and their frequencies, sorted by frequency
  my @heap;
  foreach my $num (keys %count) {
    push @heap, [$num, $count{$num}];
  }
  @heap = sort { $a->[1] <=> $b->[1] } @heap;

  # Pop the top k elements off the heap
  my @result;
  while (@result < $k) {
    push @result, shift @heap;
  }

  # Sort the resulting array by frequency in descending order
  @result = sort { $b->[1] <=> $a->[1] } @result;

  # Return the elements
  return map { $_->[0] } @result;
}

# Unit tests
my @tests = (
  # Test empty array
  {
    nums => [],
    k => 5,
    expected => [],
  },
  # Test array with fewer elements than k
  {
    nums => [1, 2, 3],
    k => 5,
    expected => [1, 2, 3],
  },
  # Test array with more elements than k
  {
    nums => [1, 2, 2, 3, 3, 3],
    k => 2,
    expected => [3, 2],
  },
 
);

foreach my $test (@tests) {
  my $result = k_most_frequent($test->{nums}, $test->{k});
  if ($result ~~ $test->{expected}) {
    print "PASS\n";
  } else {
    print "FAIL\n";
  }
}

File: topk.go
package main

import (
	"fmt"
	"sort"
)

// Returns the k most frequent elements in the given slice
func kMostFrequent(nums []int, k int) []int {
	// Return an empty slice if the input slice is empty
	if len(nums) == 0 {
		return []int{}
	}

	// Return the entire slice if k is greater than the number of elements
	if k > len(nums) {
		return nums
	}

	// Use a map to count the frequency of each element
	count := make(map[int]int)
	for _, num := range nums {
		count[num]++
	}

	// Create a min-heap of elements and their frequencies, sorted by frequency
	heap := make([][]int, 0, len(count))
	for num, freq := range count {
		heap = append(heap, []int{num, freq})
	}
	sort.Slice(heap, func(i, j int) bool { return heap[i][1] < heap[j][1] })

	// Pop the top k elements off the heap
	result := make([]int, 0, k)
	for len(result) < k {
		result = append(result, heap[0][0])
		heap = heap[1:]
	}

	// Sort the resulting slice by frequency in descending order
	sort.Slice(result, func(i, j int) bool { return count[result[i]] > count[result[j]] })

	return result
}

func main() {
	// Unit tests
	tests := []struct {
		nums   []int
		k      int
		expect []int
	}{
		// Test empty slice
		{
			nums:   []int{},
			k:      5,
			expect: []int{},
		},
		// Test slice with fewer elements than k
		{
			nums:   []int{1, 2, 3},
			k:      5,
			expect: []int{1, 2, 3},
		},
		// Test slice with more elements than k
		{
			nums:   []int{1, 2, 2, 3, 3, 3},
			k:      2,
			expect: []int{3, 2},
		},
		// Test slice with all elements the same
		{
			nums:   []int{1, 1, 1, 1, 1},
			k:      3,
			expect: []int{1},
		},
	}

	for _, test := range tests {
		result := kMostFrequent(test.nums, test.k)
		if fmt.Sprint(result) == fmt.Sprint(test.expect) {
			fmt.Println("PASS")
		} else {
			fmt.Println("FAIL")
		}
	}
}
